### PR TITLE
Bugfix optimal_edit_paths + randomized test

### LINKED ISSUE #8099: nx.optimal_edit_paths not finding optimal
This substition cost function is triggering a corner case for the algorithm where it detects it has the same "cost" to perform an insertion operation than edit.

Here is a minimum code snippet that reproduces the issue:

```python
import networkx as nx


def node_subst_cost(node_1:dict, node_2:dict)-> float:
    if node_2['i'] == 1:
        return 0
    elif node_2['i'] == 2 or node_1['i'] == 3:
        return 2
    else:
        return 2.2


def clique(nodes):
    G = nx.Graph()
    for x in nodes:
        G.add_node(x, i=x)
        for y in nodes:
            if x != y:
                G.add_edge(x, y)
    return G


example_graphs = [
    clique([1,2,3]),
    clique([3,2,1])
]

for eg in example_graphs:
    edits, distance = nx.optimal_edit_paths(eg, example_graphs[0], node_subst_cost=node_subst_cost)
    print(distance)
    for x in edits:
        for y in x:
            print(y)
        print('-')
    print('______')
```

Which outputs:
```
8.0
[(1, None), (2, 1), (3, 3), (None, 2)]
[((1, 2), None), ((1, 3), None), ((2, 3), (1, 3)), (None, (1, 2)), (None, (2, 3))]
-
______
4.0
[(3, 3), (2, 1), (1, 2)]
[((3, 2), (1, 3)), ((3, 1), (2, 3)), ((2, 1), (1, 2))]
-
______
```

The algorithm is internally computing the edit distance matrix and then calling https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linear_sum_assignment.html to detect the best step. We can see that equivalent matrices are leading to different assignments with the same cost:

```
make_CostMatrix [[ 0.   2.   2.2  1.  19.6 19.6]
 [ 0.   2.   2.2 19.6  1.  19.6]
 [ 0.   2.2  2.  19.6 19.6  1. ]
 [ 1.  19.6 19.6  0.   0.   0. ]
 [19.6  1.  19.6  0.   0.   0. ]
 [19.6 19.6  1.   0.   0.   0. ]]

assignment = [0 1 2 5 3 4] [2 0 1 3 4 5]
assignment cost = 4.0
```
make_CostMatrix [[ 0.   2.2  2.   1.  19.6 19.6]
 [ 0.   2.   2.2 19.6  1.  19.6]
 [ 0.   2.   2.2 19.6 19.6  1. ]
 [ 1.  19.6 19.6  0.   0.   0. ]
 [19.6  1.  19.6  0.   0.   0. ]
 [19.6 19.6  1.   0.   0.   0. ]] 

assignment = [0 1 2 5 3 4] [2 0 1 3 4 5]
assignment cost = 4.0
```

Algorithm is getting confused because it only costs 1 to insert a new node, I believe the bug is when handling insertions: instead of assigning a node of each graph, the algorithm requires another iteration to do that (which ends up increasing the cost).

### PR DESCRIPTION
**Pull request by amcandio**
Fixes https://github.com/networkx/networkx/issues/8099.

Revisiting https://github.com/networkx/networkx/pull/8421 by @VanitasCodes. I simplified the tests so only one small randomized test is enough to uncover the issue.

The test is based on generating random cost/edge substitutions and using a dummy brute force implementation as baseline. The dummy implementation is only based on performing node and edge substitutions so it is guaranteed to find an upper bound. On its current version it generates 10 test cases and it takes ~2.3seconds to run. I tested locally with more examples and all of them passsed.

```
================================================================================================ test session starts =================================================================================================
platform darwin -- Python 3.13.5, pytest-8.4.1, pluggy-1.6.0
rootdir: /Users/candioti/workplace/networkx
configfile: pyproject.toml
collected 9491 items / 9481 deselected / 3 skipped / 10 selected                                                                                                                                                     

networkx/algorithms/tests/test_similarity.py ..........                                                                                                                                                        [100%]

=================================================================================== 10 passed, 3 skipped, 9481 deselected in 2.34s ===================================================================================

```

### CODE REVIEW

File: networkx/algorithms/tests/test_similarity.py [hunk 1]
```diff
@@ -1156,3 +1156,109 @@ def test_panther_similarity_k_parameter_returns_k_results(self):
...
+                            H.edges[mapping[u_idx], mapping[v_idx]]
+                            if H.has_edge(mapping[u_idx], mapping[v_idx])
+                            else None
+                        )
+                        total += edge_cost(e1, e2)
+
+            if total < min_cost:
+                min_cost = total
+
+        return min_cost
```
Comment (dschult) [inline networkx/algorithms/tests/test_similarity.py:1189]:
```python
    def naive_ged_upper_bound(self, G, H, node_cost, edge_cost):
        """Brute-force GED upper over all node permutations.

        This method does not evaluate deletion nor additions,
        so it might not find the optimal solution.
        """
        min_cost = float("inf")

        for listing in itertools.permutations(nodes_H, len(G)):
            mapping = dict(zip(G, listing))
            total = sum(node_cost(u, v) for u, v in mapping.items())
            for u in G:
                if total > min_cost:
                    break
                for v in G:
                    if u < v:
                        Ge = (u, v) if G.has_edge(u, v) else None
                        He = mapping[u], mapping[v]
                        He = He if H.has_edge(*He) else None
                        total += edge_cost(Ge, He)

            if total < min_cost:
                min_cost = total

        return min_cost
```

File: networkx/algorithms/tests/test_similarity.py [hunk 2]
```diff
@@ -1156,3 +1156,109 @@ def test_panther_similarity_k_parameter_returns_k_results(self):
...
+        node_cost = self.make_random_node_cost(G, H, rng)
+        edge_cost = self.make_random_edge_cost(G, H, rng)
+
+        _, ged_dist = nx.optimal_edit_paths(
+            G, H, node_subst_cost=node_cost, edge_subst_cost=edge_cost
+        )
+
+        upper_bound = self.naive_ged_upper_bound(G, H, node_cost, edge_cost)
+
+        assert ged_dist <= upper_bound + 1e-8
```
Comment (dschult) [inline networkx/algorithms/tests/test_similarity.py:1215]: Seems like we don't need all these indexes for this test. And even if we do, we can set/get them in a slimmer way.  Does this do what we want here without the indexes and helper functions?
```python
    def test_optimal_edit_paths_random_node_and_edge_substitutions(self, size, seed):
        rng = random.Random(seed)

        # Generate two complete graphs
        G = nx.complete_graph(size)
        H = nx.complete_graph(size)

        # Random node/edge substitution cost
        max_cost = max(len(G), len(H))
        node_values = {(u, v): max_cost * rng.random() for u in G for v in H}
        node_cost = node_values.get
        Ges = list(G.edges()) + [None]
        Hes = list(G.edges()) + [None]
        edge_values = {(Ge, He): max_cost * rng.random() for Ge in Ges for He in Hes}
        edge_cost = edge_values.get

        _, ged_dist = nx.optimal_edit_paths(
            G, H, node_subst_cost=node_cost, edge_subst_cost=edge_cost
        )

        upper_bound = self.naive_ged_upper_bound(G, H, node_cost, edge_cost)

        assert ged_dist <= upper_bound + 1e-8
```

**Review by dschult — COMMENTED** [review]
The fix here involves yielding all possible "linear sum assignments" of the nodes sorted by increasing cost estimate instead of only the first one. This looks like a bug that should be fixed. :)

Thanks @amcandio for narrowing the tests to a single test. I think the naive upper bound function is a great idea. Glad it actually identifies the problematic cases.

I think the index handling of nodes and edges is not needed in this test and that removes all the other test helper functions. In the end this test class would only have the naive function and the test itself as methods. Maybe they could just be functions instead of a class, but I'm fine either way. The goal is to make the test simpler.
Warning: I haven't run this suggested code. :) 

Does this do what we need here?

**Comment by amcandio** [comment]
Thanks Dan! Let me try to simplify this test during the weekend

**Comment by amcandio** [comment]
@dschult the indices are needed because the cost function receive data (that's a limitation of the interface), but I was still able to simplify the test logic significantly. Let me know what you think!

File: networkx/algorithms/tests/test_similarity.py [hunk 3]
```diff
@@ -233,7 +233,7 @@ def test_optimal_edit_paths(self):
         G2 = cycle_graph(3)
         paths, cost = optimal_edit_paths(G1, G2)
         assert cost == 1
-        assert len(paths) == 6
+        assert len(paths) == 15
```
Comment (dschult) [inline networkx/algorithms/tests/test_similarity.py:236]: This looks very good!  

I was stumped for a while on the number of paths becoming bigger -- until I figured out that this is the number of paths, not the length of paths. :}  Now I'm stumped on another aspect that shows my lack of understanding about edit paths...


When I manually type this short test in, I get 6 paths (on `main`)and 15 paths (on PR), just as indicated in `test_optimal_edit_paths`.
<details> <summary> code to compare main and PR </summary>

```python
G1 = nx.path_graph(3)
G2 = nx.cycle_graph(3)
paths, cost = nx.optimal_edit_paths(G1, G2)
assert cost == 1
for p in paths:
    print(p)
```
</details>

But when I look at the paths themselves, I get confused.  Some of the 6 paths are not in the list of 15.  

For example, this path appears with the `main` version. It doesn't show in the paths of the current PR.
```
([(0, 0), (2, 1), (1, 2)], [(None, (0, 1)), ((0, 1), (0, 2)), ((1, 2), (1, 2))])
```
But there are other paths that are just a rearrangement of this path, e.g.
```
([(1, 2), (0, 0), (2, 1)], [((0, 1), (0, 2)), ((1, 2), (1, 2)), (None, (0, 1))])
```
So, maybe those are counted as "the same path"...  But then inside the 15 paths there are other rearrangements that seem to count as different paths.
```
([(1, 2), (0, 0), (2, 1)], [((0, 1), (0, 2)), ((1, 2), (1, 2)), (None, (0, 1))])
([(1, 2), (2, 1), (0, 0)], [((1, 2), (1, 2)), ((0, 1), (0, 2)), (None, (0, 1))])
```

How can we tell when the path count is correct? That is, what makes two paths different or the same?

<details> <summary> Full comparison </summary>

Full list of 6 paths from main (only the 1st is in the list of 15)
```
([(0, 0), (1, 1), (2, 2)], [((0, 1), (0, 1)), ((1, 2), (1, 2)), (None, (0, 2))])
([(0, 0), (2, 1), (1, 2)], [(None, (0, 1)), ((0, 1), (0, 2)), ((1, 2), (1, 2))])
([(1, 0), (0, 1), (2, 2)], [((0, 1), (0, 1)), ((1, 2), (0, 2)), (None, (1, 2))])
([(1, 0), (2, 1), (0, 2)], [((1, 2), (0, 1)), ((0, 1), (0, 2)), (None, (1, 2))])
([(2, 0), (0, 1), (1, 2)], [(None, (0, 1)), ((0, 1), (1, 2)), ((1, 2), (0, 2))])
([(2, 0), (1, 1), (0, 2)], [((1, 2), (0, 1)), ((0, 1), (1, 2)), (None, (0, 2))])
```

Full list of 15 paths found by the PR
```
([(0, 0), (1, 1), (2, 2)], [((0, 1), (0, 1)), ((1, 2), (1, 2)), (None, (0, 2))])
([(0, 0), (2, 2), (1, 1)], [(None, (0, 2)), ((0, 1), (0, 1)), ((1, 2), (1, 2))])
([(0, 0), (1, 2), (2, 1)], [((0, 1), (0, 2)), ((1, 2), (1, 2)), (None, (0, 1))])
([(1, 1), (0, 0), (2, 2)], [((0, 1), (0, 1)), ((1, 2), (1, 2)), (None, (0, 2))])
([(1, 1), (2, 2), (0, 0)], [((1, 2), (1, 2)), ((0, 1), (0, 1)), (None, (0, 2))])
([(1, 1), (0, 2), (2, 0)], [((0, 1), (1, 2)), ((1, 2), (0, 1)), (None, (0, 2))])
([(2, 2), (0, 0), (1, 1)], [(None, (0, 2)), ((0, 1), (0, 1)), ((1, 2), (1, 2))])
([(2, 2), (1, 1), (0, 0)], [((1, 2), (1, 2)), ((0, 1), (0, 1)), (None, (0, 2))])
([(2, 2), (0, 1), (1, 0)], [(None, (1, 2)), ((0, 1), (0, 1)), ((1, 2), (0, 2))])
([(0, 2), (1, 0), (2, 1)], [((0, 1), (0, 2)), ((1, 2), (0, 1)), (None, (1, 2))])
([(0, 2), (2, 1), (1, 0)], [(None, (1, 2)), ((0, 1), (0, 2)), ((1, 2), (0, 1))])
([(0, 2), (1, 1), (2, 0)], [((0, 1), (1, 2)), ((1, 2), (0, 1)), (None, (0, 2))])
([(1, 2), (0, 0), (2, 1)], [((0, 1), (0, 2)), ((1, 2), (1, 2)), (None, (0, 1))])
([(1, 2), (2, 1), (0, 0)], [((1, 2), (1, 2)), ((0, 1), (0, 2)), (None, (0, 1))])
([(1, 2), (0, 1), (2, 0)], [((0, 1), (1, 2)), ((1, 2), (0, 2)), (None, (0, 1))])
```
</details>

Comment (amcandio) [reply networkx/algorithms/tests/test_similarity.py:236]: This code is very complicated, but from my understanding, not having all paths could be explained by the pruning step. So we might stop visiting some paths after funding others.

Is this method supposed to be returning all paths?

Comment (dschult) [reply networkx/algorithms/tests/test_similarity.py:236]: The docs say "Returns all minimum-cost edit paths transforming G1 to G2."

But the word "all" in math is too often used to mean "all, up to equivalency". 
I understand that order can matter for these paths -- if you change a node before you change an edge on that node, the path will look different.  But it also seems clear that order sometimes doesn't matter -- if the node change happens in a part of the graph that isn't close to the edge change.
And I'm not sure what the aim of this function is... The paper talks mostly about finding whether a path exists or not.  I haven't found a part that cares about how many paths tie for the best distance.

Comment (amcandio) [reply networkx/algorithms/tests/test_similarity.py:236]: I'm starting to think this method of returning all paths doesn't make much sense.

Take two complete graphs G1 and G2 of size N, where the cost of converting from one node to the other is 1. There should be N! different canonical paths.

Do we need a method to return all these paths?

Comment (rossbar) [reply networkx/algorithms/tests/test_similarity.py:236]: The simple C3/P3 example is very helpful, thanks for putting this up @dschult !

To me anyways, the 6 results (i.e. the output of main) is a better outcome than the 15. The main reason being that the 15 outputs just repeat the same solution (in a different order). For example, compare the first and second rows in the table of 15 paths. Note how the values inside the lists are identical; it's only the ordering of the items inside the list that are different.

These don't strike me as distinct solutions... unless we consider the order of the outputs as indicative of the "order" of the edits (i.e. interpret the sequence as a path in the networkx-sense)[^1]. If that's the case, then we're missing even more solutions for all the possible permutations of operation order (to be clear - I don't think this is the correct interpretation!).

Ultimately then, the increase in the number of paths from 6 -> 15 does not seem to me to be a step in the right direction. That's not to say there isn't a bug though. There are many different potential bugs, but to me the most serious would be missing a lower-cost optimal path as opposed to not returning all possible path with a given cost. If we could find a single instance where that's happening, that'd really help guide what to do next!

---

As an aside, I used the C3/P3 example to ask the question: how do I get what I expect? In other words - how do I prevent all of the solutions that involve swapping node labels to get the one and only answer that involves adding the edge to close the cycle. Here's the simplest recipe I could come up with to do so:

```python
>>> G = nx.cycle_graph(3)
>>> H = nx.path_graph(3)
# Store the node labels as attributes too so they are accessible
# to the ins/del/subst callables
>>> nx.set_node_attributes(G, {n: n for n in G}, name="label")
>>> nx.set_node_attributes(H, {n: n for n in H}, name="label")

>>> def node_subst_cost(n_g, n_h):
...     if n_g["label"] == n_h["label"]:
...         return 0
...     return 1

>>> nx.optimal_edit_paths(G, H, node_subst_cost=node_subst_cost)
([([(0, 0), (1, 1), (2, 2)],
   [((0, 1), (0, 1)), ((0, 2), (0, 2)), ((1, 2), None)])],
 1.0)
```

When I run this on this PR, I get the 6 results instead of 1.

[^1]: that gets even more complicated given that we treat the node edits and edge edits separately.

**Comment by dschult** [comment]
I think the way to void returning paths which are simple reorderings of the edits is to keep track of ties and when a tie is encountered use a global node ordering to pick the next node. This is how ISMAGS yields isomorphisms without reporting symmetrically equivalent mappings. So its not surprising to see that used here too. replacing the `min` with all possible tie results undoes that.

Should we close this PR then? Or are there other useful things here...

File: networkx/algorithms/similarity.py [hunk 4]
```diff
@@ -866,30 +870,37 @@ def get_edit_ops(
...
                 reduce_C(Cv.C, (i,), (j,), m, n),
                 reduce_ind(Cv.lsa_row_ind, (i, m + j)),
                 reduce_ind(Cv.lsa_col_ind, (j, n + i)),
                 Cv.ls - Cv.C[i, j],
             )
-            yield (i, j), Cv_ij, xy, Ce_xy, Cv.C[i, j] + localCe.ls
+            first_candidates.append(((i, j), Cv_ij, xy, Ce_xy, Cv.C[i, j] + localCe.ls))
+
+        # Sort by total cost estimate and yield most promising first
+        yield from sorted(first_candidates, key=lambda t: t[4] + t[1].ls + t[3].ls)
```
Comment (dschult) [inline networkx/algorithms/similarity.py:903]: Can we just yield the minimum over all ties here? 
That will keep us from choosing the node before considering costs for the edges. And yet we still provide only one path from any remaining "ties".

```
        # Sort by total cost estimate and yield most promising first
        yield yield min(first_candidates, key=lambda t: t[4] + t[1].ls + t[3].ls)
```

Then if that works, we can go back and have the loop only compute the min and not store the whole list along the way.

**Comment by Peiffap** [comment]
I found a smaller reproducer for the non-optimal-cost behavior, using only positive costs:

```python
import networkx as nx

G = nx.Graph()
G.add_node(1, i=1)
G.add_node(2, i=2)
G.add_edge(1, 2)
H = G.copy()


def node_subst_cost(a, b):
    return 1 if b["i"] == 1 else 2


def edge_subst_cost(a, b):
    return 1


paths, distance = nx.optimal_edit_paths(
    G,
    H,
    node_subst_cost=node_subst_cost,
    edge_subst_cost=edge_subst_cost,
)

print(distance)
print(paths[0])
```

On `main`, this gives:

```python
5.0
([(1, 1), (2, None), (None, 2)], [((1, 2), None), (None, (1, 2))])
```

But there is a valid path with cost `4`:

```python
node_path = [(1, 1), (2, 2)]
edge_path = [((1, 2), (1, 2))]
```

That costs `1 + 2 + 1 = 4`. So the bug seems to be that a node-level tie between substitution and delete+insert is resolved before considering the edge cost fallout.

This PR returns the correct distance `4.0` for this example, though it returns three tied paths. Dan’s `min(first_candidates, ...)` direction looks like the right way to keep the cost fix without returning every tied first candidate, but I think the patch should bind the selected candidate explicitly. The code below still uses `fixed_i, fixed_j = i, j`, and after looping over the first candidates those names may refer to the last candidate considered rather than the one selected by `min`.

Something like this seems safer:

```python
first = min(first_candidates, key=lambda t: t[4] + t[1].ls + t[3].ls)
yield first
fixed_i, fixed_j = first[0]
```

**Comment by amcandio** [comment]
Sorry folks, I haven't had much time to follow up on this lately!

> Can we just yield the minimum over all ties here? That will keep us from choosing the node before considering costs for the edges. And yet we still provide only one path from any remaining "ties".
> 
> ```
>         # Sort by total cost estimate and yield most promising first
>         yield yield min(first_candidates, key=lambda t: t[4] + t[1].ls + t[3].ls)
> ```
> 
> Then if that works, we can go back and have the loop only compute the min and not store the whole list along the way.

Applying that makes sense but then this tests starts failing due after returning 24 as the edit distance:
```
    def test_optimize_graph_edit_distance(self):
        G1 = circular_ladder_graph(2)
        G2 = circular_ladder_graph(6)
        bestcost = 1000
        for cost in optimize_graph_edit_distance(G1, G2):
            assert cost < bestcost
            bestcost = min(bestcost, cost)
        assert bestcost == 22
```

This is an interesting observation:
```
 The code below still uses fixed_i, fixed_j = i, j, and after looping over the first candidates those names may refer to the last candidate considered rather than the one selected by min.

Something like this seems safer:

first = min(first_candidates, key=lambda t: t[4] + t[1].ls + t[3].ls)
yield first
fixed_i, fixed_j = first[0]
```

But when replacing that by first, the randomized test `test_optimal_edit_paths_random_node_and_edge_substitutions`  is finding counter-examples for when the distance is not optimal. I couldn't get to modify current algo to produce non-duplicate paths (without breakign optimality in some examples).

**Comment by rossbar** [comment]
Thanks a ton @Peiffap for sharing the reproducer!

I tried to dive in to see if I couldn't understand the example better. One thing that's always bugged me about the interface here is what happens when only a subset of the `_del`, `_ins`, `_subst` cost fns are defined --- i.e. what are the defaults used by the other ones? I think the natural assumption is that all non-explicitly-defined operations default to 1, but I don't know that that's the case.

Starting with your example, I see the same behavior on main:

```python
# Problem setup
>>> G = nx.Graph()
>>> G.add_node(1, wt=1)
>>> G.add_node(2, wt=2)
>>> G.add_edge(1, 2)
>>> H = G.copy()
>>> def node_sub_cost(u, v):
...     return v.get("wt", 1)
>>> def edge_sub_cost(u, v):
...     return 1

# Reproduce
>>> nx.optimal_edit_paths(
...     G, H, node_subst_cost=node_sub_cost, edge_subst_cost=edge_sub_cost
... )
([([(1, 1), (2, None), (None, 2)], [((1, 2), None), (None, (1, 2))]),
  ([(2, 1), (1, None), (None, 2)], [((1, 2), None), (None, (1, 2))])],
 5.0)
```

This is, I gather, what you were seeing as well (+ one extra possibility due to using the node attrs directly instead of special-casing one of them).

In fact you get the same result without explicitly defining any of the edge operations:

```python
>>> nx.optimal_edit_paths(G, H, node_subst_cost=node_sub_cost)
([([(1, 1), (2, None), (None, 2)], [((1, 2), None), (None, (1, 2))]),
  ([(2, 1), (1, None), (None, 2)], [((1, 2), None), (None, (1, 2))])],
 5.0)
```

In my head, this is equivalent to the reproducer above.

Now, what I wanted to check was what happens when the user is explicit about *all* operation costs rather than defining only a subset (e.g. substitution-only, as above). If we enforce that deletions/substitutions reflect the node weight (and set all edge operations to unity, as is implied by the fact there are no edge attributes), then I think we get the "expected" answer:

```python
# Explicit node operation costs
>>> def node_ins_cost(u):
...     return u.get("wt", 1)
>>> def node_del_cost(v):
...     return v.get("wt", 1)
# And edge operations
>>> def edge_ins_cost(e):
...     return 1
>>> def edge_del_cost(e):
...     return 1
```

Now, with all costs explicitly defined:

```python
>>> nx.optimal_edit_paths(
...     G,
...     H,
...     node_subst_cost=node_sub_cost,
...     node_ins_cost=node_ins_cost,
...     node_del_cost=node_del_cost,
...     edge_subst_cost=edge_subst_cost,
...     edge_ins_cost=edge_ins_cost,
...     edge_del_cost=edge_del_cost,
... )
([([(1, 2), (2, 1)], [((1, 2), (1, 2))]),
  ([(2, 2), (1, 1)], [((1, 2), (1, 2))])],
 4.0)
```

... the correct lowest-cost path(s)

---

So - my question is: is there a problem with the algorithm, or does the issue perhaps lie with the default cost fns when only a subset of the costs are explicitly defined? If the latter, then the good news is it should be an easy fix to put guards against this; i.e. if you define any of `_ins`, `_del` and/or `_subst` you must define all 3.
